src/analysis/jump.ml

```diff
@@ -127,43 +127,101 @@ let rec skip_non_moving pos = function
-        | Some (_, f) -> f
-        | None -> raise (No_predicate target)
-      )
+    let nodes = skip_non_moving pos enclosings in
+    let target = ["match"] in
+    let pred =
+      List.map target ~f:(fun target ->
+          match List.find_some m_pred ~f:(fun (name, _) -> name = target) with
+          | Some (_, f) -> f
+          | None -> raise (No_predicate target)
+        )
     in
-    if String.length target = 0 then
-      `Error "Specify target"
-    else begin
-      let nodes = skip_non_moving pos enclosings in
-      let node = find_node preds nodes in
-      let node_loc = Browse_raw.node_real_loc Location.none node in
-      `Found node_loc.Location.loc_start
-    end
+    let node = find_node pred nodes in
+    (* Check if the found node is a match predicate *)
+    match match_pred [node] with
+    | Some (Expression { exp_desc = Texp_match (_, cases, _); _ }) ->
+      let case = find_case_pos pos cases cmp_fn in      
+      (match case with
+      | Some location ->
+        `Found location
+      | None -> (match match_target with 
+            | "match-next-case" -> `Error "No next case found"
+            | "match-prev-case" -> `Error "No previous case found"
+            | _ -> raise (No_matching_target)))
+    | _ ->
+      raise (No_matching_target)
   with
+  | No_matching_target ->
+    `Error "Cursor is not in a match predicate"
   | No_predicate target ->
     `Error ("No predicate for " ^ target)
-  | No_matching_target ->
-    `Error "No matching target"
 
+let get_match_next_case typed_tree pos target =
+  get_match_case_pos typed_tree pos (fun pos case_loc -> pos.Lexing.pos_cnum < case_loc.pos_cnum) target
+
+let get_match_prev_case typed_tree pos target =
+  get_match_case_pos typed_tree pos (fun pos case_loc -> pos.Lexing.pos_cnum > case_loc.pos_cnum) target
+
+let get typed_tree pos target =
+  match target with
+  | "match-next-case" -> get_match_next_case typed_tree pos target
+  | "match-prev-case" -> get_match_prev_case typed_tree pos target
+  | _ ->
+    let roots = Mbrowse.of_typedtree typed_tree in
+    let enclosings =
+      match Mbrowse.enclosing pos [roots] with
+      | [] -> []
+      | l -> List.map ~f:snd l
+    in
+
+    let all_preds = [
+      "fun", fun_pred;
+      "let", let_pred;
+      "module", module_pred;
+      "match", match_pred;
+    ] in
+    let targets = Str.split (Str.regexp "[, ]") target in
+    try
+      let preds =
+        List.map targets ~f:(fun target ->
+          match List.find_some all_preds ~f:(fun (name, _) -> name = target) with
+          | Some (_, f) -> f
+          | None -> raise (No_predicate target)
+        )
+      in
+      if String.length target = 0 then
+        `Error "Specify target"
+      else begin
+        let nodes = skip_non_moving pos enclosings in
+        let node = find_node preds nodes in
+        let node_loc = Browse_raw.node_real_loc Location.none node in
```

Comment:
Here, if we found a match node and the target was a case, we switch to the new logic.

It's a shame that we have to break the elegant implementation. I wonder if Merlin already has a command to move to siblings and not only ancestors.